aten/src/ATen/NestedTensorImpl.cpp

```diff
@@ -108,43 +108,62 @@ c10::DispatchKeySet generate_nested_key_set(at::Tensor buffer) {
 }
 
 NestedTensorImpl::NestedTensorImpl(
-    at::Tensor buffer,
+    int64_t buffer_size,
+    Storage storage,
+    c10::DispatchKeySet key_set,
+    const caffe2::TypeMeta data_type,
     at::Tensor nested_size_tensor,
     at::Tensor nested_stride_tensor,
     std::vector<int64_t> offsets)
-    : TensorImpl(
-          Storage(buffer.storage()),
-          generate_nested_key_set(buffer),
-          buffer.dtype()),
+    : TensorImpl(std::move(storage), key_set, data_type),
+      buffer_size_(buffer_size),
       nested_size_tensor_(std::move(nested_size_tensor)),
       nested_stride_tensor_(std::move(nested_stride_tensor)),
       offsets_(std::move(offsets)),
-      opt_sizes_(construct_opt_sizes(nested_size_tensor_))
-{
-  auto buffer_size_vec{buffer.unsafeGetTensorImpl()->sizes()};
-  TORCH_INTERNAL_ASSERT(
-      buffer_size_vec.size() == 1,
-      "NestedTensorImpl buffer is required to be 1 dimensional but got a buffer with ",
-      buffer.dim(),
-      " dimensions.");
-  buffer_size_ = buffer_size_vec[0];
-
+      opt_sizes_(construct_opt_sizes(nested_size_tensor_)) {
   TORCH_WARN_ONCE(
       "The PyTorch API of nested tensors is in prototype stage and will change "
       "in the near future.");
-  TORCH_INTERNAL_ASSERT(buffer.is_cuda() || buffer.is_cpu(), "NestedTensorImpl buffer must be either CUDA or CPU but got ", buffer.device());
```

Comment:
You can still check the storage device if you want to preserve this.

Reply:
Will add those checks